File: src/Lexer.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}

module Lexer where

import Data.Char
import Debug.Trace
import Help
import Token

scan :: String -> [Token]
scan [] = []
scan source = help [] source
  where
    help :: [Token] -> String -> [Token]
    help tokens "" = reverse $ Token EOF "eof" None : tokens
    help tokens (':' : '=' : ode) = help (Token BIND ":=" InfixOp : tokens) ode
    help tokens code@(c : ode) =
      case c of
        '+' -> help (Token PLUS "+" InfixOp : tokens) ode
        '*' -> help (Token STAR "*" InfixOp : tokens) ode
        '^' -> help (Token POWER "^" InfixOp : tokens) ode
        '|' -> help (Token VBAR "|" InfixOp : tokens) ode
        ',' -> help (Token COMMA "," InfixOp : tokens) ode
        '\n' -> help (simpleToken SEMICOLON "\n" : tokens) ode
        '-' ->
          if head ode == '>'
            then help (simpleToken LAMBDA "->" : tokens) $ tail ode
            else help (Token MINUS "-" InfixOp : tokens) ode
        _
          | c == '/' || c == '\\' || c == '!' ->
            if head ode == '='
              then help (Token NEQUAL "/=" InfixOp : tokens) $ slice ode 1 $ length ode
              else help (Token SLASH "/" InfixOp : tokens) ode
        _ | c == '(' -> help (simpleToken LPAREN "(" : tokens) ode
        _ | c == ')' -> help (simpleToken RPAREN ")" : tokens) ode
        _ | c == '[' -> help (simpleToken LCPAREN "[" : tokens) ode
        _ | c == ']' -> help (simpleToken RCPAREN "]" : tokens) ode
        _ | c == '{' -> help (simpleToken LBRACE "{" : tokens) ode
        _ | c == '}' -> help (simpleToken RBRACE "}" : tokens) ode
        _ | c == '=' || c == '<' || c == '>' -> addEqual code
        _ | isNumber c -> addNum code False ""
        _ | isLetter c -> addName code
        _ | c == ' ' || c == '\t' || c == '\r' -> help tokens ode
        _ | otherwise -> error $ "Syntax error: " ++ show c
      where
        simpleToken :: TokenType -> String -> Token
        simpleToken token content = Token token content None

        addEqual :: String -> [Token]
        addEqual ('<' : '=' : xs) = help (Token LEQUAL "<=" InfixOp : tokens) xs
        addEqual ('>' : '=' : xs) = help (Token GEQUAL ">=" InfixOp : tokens) xs
        addEqual ('=' : '=' : xs) = help (Token DEQUAL "==" InfixOp : tokens) xs
        addEqual ('<' : xs) = help (Token LESS "<" InfixOp : tokens) ode
        addEqual ('>' : xs) = help (Token GREAT ">" InfixOp : tokens) ode
        addEqual ('=' : xs) = help (Token EQUAL "=" InfixOp : tokens) ode

        addNum :: String -> Bool -> String -> [Token]
        addNum ('.' : xs) isFloat res = addNum xs True (res ++ ".")
        addNum xs@(x : s) isFloat res
          | isNumber x = addNum s isFloat (res ++ [x])
          | otherwise = resultNum isFloat res xs
        addNum [] isFloat res = resultNum isFloat res []

        resultNum :: Bool -> String -> String -> [Token]
        resultNum True res next = help (Token FLOAT res Num : tokens) next
        resultNum False res next = help (Token INT res Num : tokens) next

        addName code = help (Token (fst4 token) (snd4 token) (thd4 token) : tokens) (fth4 token)
          where
            token = case code of
              ('i' : 'f' : xs) -> (IF, "if", Name, xs)
              ('t' : 'h' : 'e' : 'n' : xs) -> (THEN, "then", None, xs)
              ('e' : 'l' : 's' : 'e' : xs) -> (ELSE, "else", None, xs)
              ('t' : 'r' : 'u' : 'e' : xs) -> (TRUE, "true", Name, xs)
              ('f' : 'a' : 'l' : 's' : 'e' : xs) -> (FALSE, "false", Name, xs)
              ('a' : 'n' : 'd' : xs) -> (AND, "and", Name, xs)
              ('o' : 'r' : xs) -> (OR, "or", Name, xs)
              ('n' : 'o' : 't' : xs) -> (NOT, "not", Name, xs)
              ('n' : 'i' : 'l' : xs) -> (NIL, "nil", Name, xs)
              xs ->
                ( NAME,
                  takeWhile isName code,
                  Name,
                  dropWhile isName xs
                )

isName :: Char -> Bool
isName c = isLetter c || isNumber c

File: Scala/src/main/scala/Lexer.scala
object Lexer {

  def scan(code: Option[String]): Option[List[Token]] = {

    def help(tokens: List[Token], start: Int): Option[List[Token]] = {

      def addNum(s: Int, r: String = "", isFloat: Boolean = false): Option[List[Token]] = {
        def result(): Option[List[Token]] = help(Token(if isFloat then Token.FLOAT
                                            else Token.INT, r.reverse) :: tokens, s)
        if s < 0 then return result()
        val c = code.get(s)
        c match {
          case '.' => addNum(s - 1, r + c, true)
          case _ => if !c.isDigit then result() else addNum(s - 1, r + c, isFloat)
        }
      }

      def addName(s: Int, r: String = ""): Option[List[Token]] = {
        def result(): Option[List[Token]] = {
          var t: TokenType = Token.NIL
          r.reverse match {
            case "if" => t = Token.IF
            case "then" => t = Token.THEN
            case "else" => t = Token.ELSE
            case "true" => t = Token.TRUE
            case "false" => t = Token.FALSE
            case "and" => t = Token.AND
            case "or" => t = Token.OR
            case "not" => t = Token.NOT
            case _ => t = Token.NAME
          }
          help(Token(t, r.reverse) :: tokens, s)
        }
        if s < 0 then return result()
        val c = code.get(s)
        if c.isLetter then addName(s - 1, r + c) else result()
      }

      def addEqual(s: Int): Option[List[Token]] = {
        def result(t: TokenType, r: String, offset: Int = 1): Option[List[Token]] =
          help(Token(t, r) :: tokens, s - offset)
        val c = code.get(s - 1)
        if (c.toString + code.get(s).toString) == ":=" then result(Token.BIND, ":=", 2)
        else c match {
          case '<' => result(Token.LEQUAL, "<=", 2)
          case '>' => result(Token.GEQUAL, ">=", 2)
          case '=' => result(Token.DEQUAL, "==", 2)
          case _ => code.get(s) match {
            case '=' => result(Token.EQUAL, "=")
            case '<' => result(Token.LESS, "<")
            case '>' => result(Token.GREAT, ">")
          }
        }
      }

      if start < 0 then return Option(tokens)
      val c = code.get(start)
      code match {
        case Some(code) => c match {
          case '+' => help(Token(Token.PLUS, "+") :: tokens, start - 1)
          case '-' => help(Token(Token.MINUS, "-") :: tokens, start - 1)
          case '*' => help(Token(Token.STAR, "*") :: tokens, start - 1)
          case '/' => help(Token(Token.SLASH, "/") :: tokens, start - 1)
          case '\\' => help(Token(Token.RSLASH, "\\") :: tokens, start - 1)
          case '!' => help(Token(Token.EXPCLAMATION, "!") :: tokens, start - 1)
          case '^' => help(Token(Token.POWER, "^") :: tokens, start - 1)
          case '(' => help(Token(Token.LPAREN, "(") :: tokens, start - 1)
          case ')' => help(Token(Token.RPAREN, ")") :: tokens, start - 1)
          case '@' => help(Token(Token.LAMBDA, "@") :: tokens, start - 1)
          case '=' | '<' | '>' | ':' => addEqual(start)
          case '\n' => help(Token(Token.SEMICOLON, "\\n") :: tokens, start - 1)
          case ' ' | '\t' | '\r' => help(tokens, start - 1)
          case _ if c.isDigit => addNum(start)
          case _ if c.isLetter => addName(start)
          case _ => Main.error("Syntax error: ", c.toString)
                    Option(tokens)
        }
        case None => None
      }
    }

    val len = code.get.length - 1
    Option(help(Nil, len).get ::: List(Token(Token.EOF, "\u0000")))
  }
}